packages/compose/src/hooks/use-debounce/index.js

```diff
@@ -0,0 +1,46 @@
+/**
+ * useDebounce hook
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect } from '@wordpress/element';
+
+/**
+ * Use to debounce a value that is changing too frequently.
+ *
+ * Example usage:
+ *
+ * ```js
+ *	// An example using the useDebounce hook with useResizeObserver()
+ *  // to debounce a rapidly changing width value
+ *
+ *	const [ resizeListener, sizes ] = useResizeObserver();
+ *
+ *	useDebounce( ( sz ) => {
+ *
+ *		// ... do your thing here ...
+ *
+ *	}, 200, sizes.width );
+ *
+ *	return (
+ *		<div>
+ *			{resizeListener}
+ *		</div>
+ *	);
+ *	```
+ *
+ * @param {Function} callback The function to call with the debounced value.
+ * @param {number}   delay    The amount to delay time in ms.
+ * @param {Array}    deps     The dependent value changing you want to debounce.
+ *
+ */
+const useDebounce = ( callback, delay, deps ) => {
+	useEffect( () => {
+		const handler = setTimeout( () => callback( deps ), delay );
```

Comment:
Is it necessary to pass the dependencies back through to the callback, if presumably the caller would already have those dependencies in scope?

Reply:
Yes, I think its needed because the setTimeout() will call the function that may need the debounced value.

Reply:
> Yes, I think its needed because the setTimeout() will call the function that may need the debounced value.

I'm not sure I understand. It would still be available in scope, wouldn't it? And if it's a point of the value being from _before_ potential changes in that value, isn't the `useEffect` handling that by cancelling and rescheduling? So we can be certain that the `deps` within the callback would be the same as the `deps` from the parent scope?

Reply:
Also, would we be wanting to pass a single array argument of the `deps` to the callback, or each dependency as a separate argument?

(Aside: Both of the above would make for good test cases)